partiql-types/src/main/java/org/partiql/value/PartiQLData.java

```diff
@@ -0,0 +1,158 @@
+package org.partiql.value;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.util.Iterator;
+import org.jetbrains.annotations.NotNull;
+import org.partiql.value.datetime.Date;
+import org.partiql.value.datetime.Time;
+import org.partiql.value.datetime.Timestamp;
+
+/**
+ * Data representing a database result set, which is usually generated by executing a statement that queries the database.
+ * <p>
+ * A {@link PartiQLData} object maintains a cursor pointing to its current position in the underlying data. Initially the
+ * cursor is positioned before the first value. The {@link #next()} method moves the cursor to the next value. Please use
+ * {@link #hasNext()} before calling {@link #next()}.
+ *
+ * @see PartiQLValueLoader#loadSingleValue(PartiQLData)
+ * @see PartiQLValue
+ */
+public interface PartiQLData extends AutoCloseable, Iterator<PartiQLValueType> {
+
+    /**
+     * Positions the reader just before the contents of the current value, which must be a container (list, bag,
+     * sexp, or struct). There's no current value immediately after stepping in, so the next thing you'll want to do is call
+     * {@link #hasNext()} and {@link #next()} to move onto the first child value.
+     * <p>
+     * If the container itself is the null value, stepIn() shall fail. Please use {@link #isNullValue()} before
+     * invoking this.
+     * <p>
+     * At any time {@link #stepOut()} may be called to move the cursor back to (just after) the parent value, even if
+     * there are more children remaining.
+     */
+    public void stepIn();
+
+    /**
+     * Positions the iterator after the current parent's value, moving up one level in the data hierarchy. There's no
+     * current value immediately after stepping out, so the next thing you'll want to do is call {@link #hasNext()} and
+     * {@link #next()} to move onto the following value.
+     */
+    public void stepOut();
```

Comment:
Why no container type arg?

Reply:
Would there be a need to pass a type? We already receive enough information from `next()`, `getType()`, and `isNullValue()` to determine whether we can/should invoke `stepIn()` and `stepOut()`.